### ISSUE TITLE
Feature request: camera elevation

### ISSUE THREAD

**Issue by IvanSanchez** [mapbox/mapbox-gl-js#3552]
## Motivation

With extruded geometries, there are use cases where the focus should be on a specific elevation.

The most obvious is  to focus on a building rooftop with a call to `Map#flyTo`. With more 3D data, users might want to have a control to travel "up" and "down" besides moving N-S-W-E.

## Design

I suggest a new property in [`CameraOptions`](https://www.mapbox.com/mapbox-gl-js/api/#CameraOptions), named `height` or `elevation`, e.g.:

```
map.flyTo({
  center: [ 63.5, 10.4 ],
  zoom: 17,
  bearing: 15,
  pitch: 45,
  elevation: 100  // assumed meters, or CRS units, as in the extrude height
})
```

### Implementation

I guess the internal camera state shall store the camera elevation somewhere, and modify the 3D transform matrix as appropriate.

This might change the viewport bounds, so the calculations for the tile ranges might change as well (my knowledge of the internals of this project is low, so I don't know if the 3D transform matrix is reused for this purpose).

Any easing function for the camera state (`flyTo` animations and the like) shall also ease and interpolate the elevation.

**Comment by 1ec5** [comment]
First of all, any `elevation` option would be mutually exclusive with the existing `zoom` option, because they’re competing ways of scaling the map. Zoom level–based APIs are appropriate for aligning with a given function stop in a style layer, whereas altitude-based APIs are appropriate for aligning to real-world features. (This is, after all, a map rather than an image viewer.)

Specifically, imagine you want to set a camera that comfortably clears the tallest building in a city. You could do it in one line by saying you want a camera with a certain altitude. Or you can guess and check until you come up with the right zoom level. But guessing and checking only works if you hard-code a specific city and a specific building.

[`flyTo`](https://www.mapbox.com/mapbox-gl-js/api/#Map#flyTo) demonstrates this point nicely. A developer can only intuit a value for the `minZoom` option if they’ve designed the map themselves and want to ensure that a certain function stop is reached. But if the goal is to simulate aircraft flying at a certain altitude – say, for a game – a [`peakAltitude`](https://www.mapbox.com/ios-sdk/api/3.3.6/Classes/MGLMapView.html#/c:objc(cs)MGLMapView(im)flyToCamera:withDuration:peakAltitude:completionHandler:) option would be far more intuitive.

The Mapbox iOS and macOS SDKs have an implementation of the requested feature in the [`altitude`](https://www.mapbox.com/ios-sdk/api/3.3.6/Classes/MGLMapCamera.html#/c:objc(cs)MGLMapCamera(py)altitude) property of MGLMapCamera. You can set the map’s zoom level, or you can set the altitude, but not both at the same time. For a tilted map, you can create the camera in two ways: either [based on the viewer’s height above the ground](https://www.mapbox.com/ios-sdk/api/3.3.6/Classes/MGLMapCamera.html#/c:objc(cs)MGLMapCamera(cm)cameraLookingAtCenterCoordinate:fromEyeCoordinate:eyeAltitude:) or [based on the center point’s viewing distance](https://www.mapbox.com/ios-sdk/api/3.3.6/Classes/MGLMapCamera.html#/c:objc(cs)MGLMapCamera(cm)cameraLookingAtCenterCoordinate:fromDistance:pitch:heading:) (since Mapbox GL doesn’t offer any way to change the focal point). Unfortunately, [the conversion between zoom levels and altitudes](https://github.com/mapbox/mapbox-gl-native/blob/6338a3e7f9aca0b771cad82ca52ebc9d6ad19871/platform/darwin/src/MGLGeometry.mm#L33-L46) is fraught, but there would be a real benefit for developers who need animations with a relation to the real world.

**Comment by ansis** [comment]
@1ec5 I agree with what you wrote but I understood the request slightly differently. You're thinking about the altitude of the camera but I think @IvanSanchez is asking about the altitude of the point the camera is looking at. This fits in which the current `CameraOptions` which describes what the camera is looking at and how, not where the camera is.

Right now you can only say "look at the point at **0m** above ground level at [-79.3, 43.6]" but the suggestion is to support saying "look at the point at **100m** above ground level at [-79.3, 43.6]". 

I think this makes sense to support.

Changing the `0` [on this line in transform.js](https://github.com/mapbox/mapbox-gl-js/blob/f1d1bf464d589632fdad3b0c1fe10f30d822fbd7/js/geo/transform.js#L455) should be what we need on the projection side.

Tile loading might automatically adapt or it might need some changes. And once that is working, if you set it to something too high for the current zoom it will try to load too many tiles. We need a temporary limit on this. The level-of-detail tile loading we need to implement for pitched view should let us remove this limit eventually.

**Comment by IvanSanchez** [comment]
> I think @IvanSanchez is asking about the altitude of the point the camera is looking at.

Yes, yes, yes, yes. Sorry if I was unclear in my first explanation.

The main use case I have in mind is being able to perform `flyTo` calls which centers the map on building rooftops, not on the building footprints, regardless of tilt and bearing.

**Comment by adamduncan** [comment]
Hey @ansis, @IvanSanchez, has there been any more exploration of this approach? Am currently working on `fitBounds`-ing some polygon extrusions with known heights, but it's all a bit chicken and egg.

**Comment by strangerintheq** [comment]
Voting up for this feature. 
In case of showing of buildings overlaying mapbox.

**Comment by andrewharvey** [comment]
I've hacked together some code to workaout the lack of core support
https://bl.ocks.org/andrewharvey/a0b56df5d605d86facc3da9723a9ef6e. It will find the corresponding map center/zoom to use to look at a point at an elevation. It's not the best code, and might not cover all cases, but it seems to work okay so far.

@1ec5 and @ansis raise good points in https://github.com/mapbox/mapbox-gl-js/issues/3552#issuecomment-260753669 and https://github.com/mapbox/mapbox-gl-js/issues/3552#issuecomment-260801024

> First of all, any elevation option would be mutually exclusive with the existing zoom option, because they’re competing ways of scaling the map. Zoom level–based APIs are appropriate for aligning with a given function stop in a style layer, whereas altitude-based APIs are appropriate for aligning to real-world features. (This is, after all, a map rather than an image viewer.)
> 
> Specifically, imagine you want to set a camera that comfortably clears the tallest building in a city. You could do it in one line by saying you want a camera with a certain altitude. Or you can guess and check until you come up with the right zoom level. But guessing and checking only works if you hard-code a specific city and a specific building.

True. My suggestion would be when passing a `[lng, lat, altitude]` to jumpTo/flyTo you would also need to pass a `cameraHeight` instead of a `zoom`. The `cameraHeight` would be how high do you want the camera when looking at `[lng, lat, altitude]`. A nice result for 45deg would be `2 x altitude`, and it would pick out the corresponding `zoom` value. Or it would be `distanceFrom` as they are equivalent, just something in real world units, not zoom levels.

> Changing the 0 on this line in transform.js should be what we need on the projection side.

Hmm, but won't that then mean that `map.getCenter()` is no longer the point in the middle of the screen?

**Comment by andrewharvey** [comment]
See also #6093

**Comment by SkylinR** [comment]
Hi :) I'm trying to display some popular buildings on map. (I need to disable by ID some default buildings and replace them with my owne models). And it's working like a charm but I get ID of that default buildings on some zoom (from 16, on higher zoom like 15.9 I cannot get ID of that single default building to replace it). So I'm there with zoom 16 and I want to display Burj Khalifa but this is quite high building and it's not visible entirely. I'm trying to change elevation to look a little bit higher then base of building to be able to display it whole on screen and there is my PROBLEM. As far as I can see all those elevation/altitude solutions works on zoom and when I'm trying to use it at the end of the day it changes map zoom (what makes that I've displayed default building because on higher zoom I cannot disable that single default building).
Is there any option to just make camera look on point like "over" map a little bit. 
Aditionally I'm rotating camera around that  point :D so it's important to not just setup one view but it must be visible from each side.

**Comment by Bulisor** [comment]
> The free camera, via [FreeCamerOptions](https://docs.mapbox.com/mapbox-gl-js/api/properties/#freecameraoptions) now allows defining the position of the camera with elevation, but the `look at point` is still based on 0 elevation.

Hey, what about `look at point`? Is there a way I can set it at a specific altitude? Thank you

**Comment by tibotiber** [comment]
Hello, I'm wondering if anything has moved on this topic of having a "target point" of the camera with an elevation. 

The gist shared by @andrewharvey is great, but if I'm not mistaken, it wouldn't solve subsequent rotation of the camera around the "desired" target point. It would rotate the camera in a way that looks inconsistent to the place the user "thinks" they're looking at. 

I believe with Mapbox Standard bringing a 3D basemap to the table, topics like this one should gain interest and not remain stale? Is Standard changing anything in how this can be approached?

Main use case for me: being able to rotate the camera around a target point located on a roof (e.g.).

**Comment by ibesora** [comment]
Hey @tibotiber , [lookAtPoint](https://docs.mapbox.com/mapbox-gl-js/api/properties/#freecameraoptions#lookatpoint) supports an `altitude` parameter that allows you to set the altitude of the focus point on the map.
Is this enough for your use case?

**Comment by tibotiber** [comment]
Thanks for pointing it out @ibesora.

I had in mind something closer to the target point of an orbit camera in 3D engines. So once you set that target point, rotation would happen around that point as anchor. 

Here I think it seems to compute a ground based coordinate to look at to "fake" the target elevation. Subsequent rotation happens around that potentially distant point, so if you're "thinking" you're focused on an elevated point, it is pretty confusing.

Am I making sense? I can also create videos to make that difference more clear if it helps. Let me know.

Potential impact I'm seeing: building a mapbox-powered indoor-based 3D experience that potentially deals with high buildings would require this.

**Comment by ibesora** [comment]
> Am I making sense?

You are, yes. Thanks for the answer
